### ISSUE TITLE
Meteor run fails if Temp directory is on drive in memory

### ISSUE THREAD

**Issue by vladimir3000** [meteor/meteor#6189]
In this case RAMDisk, log follows:

```
C:\MProjects\leaderboard>meteor run >1.txt

C:\Users\Vladimir\AppData\Local\.meteor\packages\meteor-tool\1.1.10\mt-os.window
s.x86_32\dev_bundle\lib\node_modules\fibers\future.js:278
                                                throw(ex);
                                                      ^
Error: EEXIST, rename 'R:\Temp\mt-uzavzq'
    at Object.Future.wait (C:\Users\Vladimir\AppData\Local\.meteor\packages\mete
or-tool\1.1.10\mt-os.windows.x86_32\dev_bundle\lib\node_modules\fibers\future.js
:398:15)
    at C:\tools\fs\files.js:1331:28
    at wrapper (C:\tools\fs\files.js:1334:20)
    at Object.files.rename (C:\tools\fs\files.js:1397:9)
    at Object.files.extractTarGz (C:\tools\fs\files.js:720:9)
    at Object.exports._extractAndConvert (C:\tools\packaging\tropohouse.js:49:9)

    at [object Object]._.extend._downloadBuildToTempDir (C:\tools\packaging\trop
ohouse.js:308:20)
    at C:\tools\packaging\tropohouse.js:483:39
    at C:\tools\utils\buildmessage.js:356:18
    at [object Object]._.extend.withValue (C:\tools\utils\fiber-helpers.js:114:1
4)
    at C:\tools\utils\buildmessage.js:349:34
    at [object Object]._.extend.withValue (C:\tools\utils\fiber-helpers.js:114:1
4)
    at C:\tools\utils\buildmessage.js:347:23
    at [object Object]._.extend.withValue (C:\tools\utils\fiber-helpers.js:114:1
4)
    at Object.enterJob (C:\tools\utils\buildmessage.js:321:26)
    at C:\tools\packaging\tropohouse.js:479:24
    at Array.forEach (native)
    at Function._.each._.forEach (C:\Users\Vladimir\AppData\Local\.meteor\packag
es\meteor-tool\1.1.10\mt-os.windows.x86_32\dev_bundle\lib\node_modules\underscor
e\underscore.js:79:11)
    at C:\tools\packaging\tropohouse.js:478:11
    at C:\tools\utils\buildmessage.js:356:18
    at [object Object]._.extend.withValue (C:\tools\utils\fiber-helpers.js:114:1
4)
    at C:\tools\utils\buildmessage.js:349:34
    at [object Object]._.extend.withValue (C:\tools\utils\fiber-helpers.js:114:1
4)
    at C:\tools\utils\buildmessage.js:347:23
    at [object Object]._.extend.withValue (C:\tools\utils\fiber-helpers.js:114:1
4)
    at Object.enterJob (C:\tools\utils\buildmessage.js:321:26)
    at Object.download (C:\tools\packaging\tropohouse.js:437:20)
    at C:\tools\packaging\tropohouse.js:575:22
    at C:\tools\utils\buildmessage.js:356:18
    at [object Object]._.extend.withValue (C:\tools\utils\fiber-helpers.js:114:1
4)
    at C:\tools\utils\buildmessage.js:349:34
    at [object Object]._.extend.withValue (C:\tools\utils\fiber-helpers.js:114:1
4)
    at C:\tools\utils\buildmessage.js:347:23
    at [object Object]._.extend.withValue (C:\tools\utils\fiber-helpers.js:114:1
4)
    at Object.enterJob (C:\tools\utils\buildmessage.js:321:26)
    at [object Object]._.extend.downloadPackagesMissingFromMap (C:\tools\packagi
ng\tropohouse.js:572:20)
    at C:\tools\project-context.js:702:25
    at C:\tools\utils\buildmessage.js:356:18
    at [object Object]._.extend.withValue (C:\tools\utils\fiber-helpers.js:114:1
4)
    at C:\tools\utils\buildmessage.js:349:34
    at [object Object]._.extend.withValue (C:\tools\utils\fiber-helpers.js:114:1
4)
    at C:\tools\utils\buildmessage.js:347:23
    at [object Object]._.extend.withValue (C:\tools\utils\fiber-helpers.js:114:1
4)
    at Object.enterJob (C:\tools\utils\buildmessage.js:321:26)
    at C:\tools\project-context.js:701:20
    at C:\tools\packaging\catalog\catalog.js:100:5
    at C:\tools\utils\buildmessage.js:268:13
    at [object Object]._.extend.withValue (C:\tools\utils\fiber-helpers.js:114:1
4)
    at C:\tools\utils\buildmessage.js:261:29
    at [object Object]._.extend.withValue (C:\tools\utils\fiber-helpers.js:114:1
4)
    at C:\tools\utils\buildmessage.js:259:18
    at [object Object]._.extend.withValue (C:\tools\utils\fiber-helpers.js:114:1
4)
    at C:\tools\utils\buildmessage.js:250:23
    at [object Object]._.extend.withValue (C:\tools\utils\fiber-helpers.js:114:1
4)
    at Object.capture (C:\tools\utils\buildmessage.js:249:19)
    at Object.catalog.runAndRetryWithRefreshIfHelpful (C:\tools\packaging\catalo
g\catalog.js:99:31)
    at ProjectContext._.extend._downloadMissingPackages (C:\tools\project-contex
t.js:700:13)
    at C:\tools\project-context.js:257:35
    at C:\tools\utils\buildmessage.js:356:18
    at [object Object]._.extend.withValue (C:\tools\utils\fiber-helpers.js:114:1
4)
    at C:\tools\utils\buildmessage.js:349:34
    at [object Object]._.extend.withValue (C:\tools\utils\fiber-helpers.js:114:1
4)
    at C:\tools\utils\buildmessage.js:347:23
    at [object Object]._.extend.withValue (C:\tools\utils\fiber-helpers.js:114:1
4)
    at Object.enterJob (C:\tools\utils\buildmessage.js:321:26)
    at ProjectContext._.extend._completeStagesThrough (C:\tools\project-context.
js:247:18)
    at ProjectContext._.extend.saveChangedMetadata (C:\tools\project-context.js:
235:10)
    at ProjectContext._.extend.prepareProjectForBuild (C:\tools\project-context.
js:240:10)
    at C:\tools\runners\run-app.js:517:29
    at C:\tools\utils\buildmessage.js:268:13
    at [object Object]._.extend.withValue (C:\tools\utils\fiber-helpers.js:114:1
4)
    at C:\tools\utils\buildmessage.js:261:29
    at [object Object]._.extend.withValue (C:\tools\utils\fiber-helpers.js:114:1
4)
    at C:\tools\utils\buildmessage.js:259:18
    at [object Object]._.extend.withValue (C:\tools\utils\fiber-helpers.js:114:1
4)
    at C:\tools\utils\buildmessage.js:250:23
    at [object Object]._.extend.withValue (C:\tools\utils\fiber-helpers.js:114:1
4)
    at Object.capture (C:\tools\utils\buildmessage.js:249:19)
    at bundleApp (C:\tools\runners\run-app.js:516:31)
    at [object Object]._.extend._runOnce (C:\tools\runners\run-app.js:594:35)
    at [object Object]._.extend._fiber (C:\tools\runners\run-app.js:858:28)
    at C:\tools\runners\run-app.js:396:12
```

**Comment by stubailo** [comment]
This seems to me like a pretty unlikely scenario since most people don't use ram disks, so I don't think we are likely to do anything about it. Some questions:

1. Why are you using a ram disk? Is it required for you for some reason?
2. What would we need to change to make it work? If it's a simple fix we might be able to accept a PR.

**Comment by stubailo** [comment]
Going to close the issue and reopen if it looks like something we might actually fix.

**Comment by williamledoux** [comment]
I am also facing this issue. To answer your questions:

1. Because it speeds up a lot of applications that writes a lot of temp files (C++ compilation for instance).
2. I don't understand what is it that meteor does that is incompatible with the folder being mounted from RAM. But if it is something hard to fix, a workaround would be to be able to specify the folder meteor is using as a temp folder.

**Comment by BadOPCode** [comment]
Stubalio this is not a unlikely scenario at all.  Majority of Linux and BSD systems that have a /tmp directory are using tmpfs which makes a RAM drive with a max allocated size of half of your total system RAM which you can actually reconfigure to do more or less.  This is why some of the people with the hanging downloading or hanging extracting say "Oh it worked but I had to wait forever"  Ya because their OS was committing sepuku trying keep up with the demands of the tmpfs wanting 100% is my guess.  This RAM /tmp drive is the way it is because when you reboot your system everything in /tmp gets wiped out.  This is great for local.socks and semaphore files. It wasn't designed for downloading and extracting files.  I *HIGHLY* recommend that you guys make your own temp directory in the .meteor path and do the exact same procedure you have now in /tmp and remove files once successfully installed.

**Comment by BadOPCode** [comment]
@williamledoux I don't believe Windows temp directory is a RAM drive like it is in *nix systems.  On *nix systems I read a blog that said to fix this Meteor issue you should set  ```export TMPDIR=$HOME/tmp```  I will confirm this later.  The only thing that makes me nervous about setting all the temp paths to a real FS is that programs like Mongo use the /tmp directory for local socks and you would lose a lot of performance taking that out of RAM storage.
Actually there is a really good chance that trying to compile while source is being temporarily stored in tmpfs will actually slow down the process.  IF the usage of the tmpfs starts to fight the compiler and OS for RAM.
It seems like the majority of Windows users are having problems with tar and 7zip being in their path is what it sounds like to me.  Or possibly path order. (The install/reinstall fiddly trick)  This could possibly effect Mac / *nix users as well.
The temporary directory running out of  RAM space can easily be figured out by ctrl+c out of the meteor process and do a ```df``` and look at the Use% of /tmp.  If it says "100%" than that is why your Meteor is hanging.

**Comment by williamledoux** [comment]
> @williamledoux I don't believe Windows temp directory is a RAM drive

@BadOPCode, Your are right it is not. But you can make it to be by using RAMDisk like the original poster and myself did. The problem is only occuring on Windows , hence @stubailo's claim that it's an unlikely scenario.

> Actually there is a really good chance that trying to compile while source is being temporarily stored in tmpfs will actually slow down the process. IF the usage of the tmpfs starts to fight the compiler and OS for RAM.

I assume that people using RAMDisk make sure to have enough RAM and even disable swap so that they get a friendly reminder in case they're running out of RAM. At least that's what I do, even if it has obvious downsizes.

**Comment by BadOPCode** [comment]
@williamledoux Actually in some cases it's happening on Linux.  More likely than not I'm going to assume it's people like myself who are running Meteor on VM's.  After getting frustrated and throwing things and having my little temper tantrum that it's all Meteor's fault... I calmed down and relooked at the issue... and it actually starts with being a NPM issue.
It's actually feasible to sit through the issue (taking FOREVER) *IF* you have enough swap for your VM.  I have tested the ```export TMPDIR="~/tmp"``` solution and it does work and it's a lot faster than trying to let swap memory resolve this.
It would be great if NPM and Meteor handled their own temporary cache.  I mean Meteor actually goes behind and removes it's temporary files already.  The hard part is done.  Just point to ~/.meteor/tmp now.  BUT NPM would still be a problem child though.
